### PR TITLE
WIP: add caching using KV

### PR DESCRIPTION
**Pull request by karthik2804**
without caching in kv
```bash
bombardier -c 1 -n 1000 http://localhost:3000/cloud/data-redis
Bombarding http://localhost:3000/cloud/data-redis with 1000 request(s) using 1 connection(s)
 1000 / 1000 [=========================================================] 100.00% 68/s 14s
Done!
Statistics        Avg      Stdev        Max
  Reqs/sec        69.01      24.34     104.92
  Latency       14.48ms     1.66ms    25.97ms
  HTTP codes:
    1xx - 0, 2xx - 1000, 3xx - 0, 4xx - 0, 5xx - 0
    others - 0
  Throughput:     2.20MB/s
```

with caching 
```bash
bombardier -c 1 -n 1000 http://localhost:3000/cloud/data-postgres
Bombarding http://localhost:3000/cloud/data-postgres with 1000 request(s) using 1 connection(s)
 1000 / 1000 [=========================================================] 100.00% 998/s 1s
Done!
Statistics        Avg      Stdev        Max
  Reqs/sec      1168.36     367.53    1642.22
  Latency        0.86ms     5.59ms   177.52ms
  HTTP codes:
    1xx - 0, 2xx - 1000, 3xx - 0, 4xx - 0, 5xx - 0
    others - 0
  Throughput:    34.10MB/s
```

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -19,7 +19,7 @@ http = "0.2.6"
 pulldown-cmark = { version = "0.9.1", default-features = false }
 serde = { version = "1.0.136", features = ["derive"] }
 serde_json = "1.0.79"
-spin-sdk = { git = "https://github.com/fermyon/spin", rev = "139c40967a75dbdd5d4da2e626d24e68f54c0a5a", optional = true }
+spin-sdk = { git = "https://github.com/fermyon/spin", optional = true }
```
Comment (itowlson) [inline Cargo.toml]: Strongly recommend `branch` or `tag` so you're not build off a moving target.

File: src/content.rs [hunk 2]
```diff
@@ -124,40 +105,19 @@ pub fn all_pages(
...
-                    // The rest of the file contains the cache contents
-                    writeln!(f, "{cache_expiry_time}")?;
-                    write!(f, "{cache_data}")?;
-                }
-                _ => {
-                    eprintln!("Cannot create Cache file");
-                }
-            }
+            store.set("bartholomew_index_cache", cache_data)?;
+            store.set("bartholomew_cache_expiry", cache_expiry_time)?;
```
Comment (itowlson) [inline src/content.rs]: Consider using shared constants for these strings, so as not to have to keep them in sync across usages.

File: src/bartholomew.rs [hunk 3]
```diff
@@ -114,10 +143,32 @@ pub fn render(req: Request) -> Result<Response> {
...
                         .unwrap_or_else(|| DEFAULT_CONTENT_TYPE.to_owned());
 
+                    let disable_page_cache = doc.head.disable_page_cache;
+
                     let data = engine
                         .render_template(doc, config, req.headers().to_owned())
                         .map_err(|e| anyhow!("Rendering {:?}: {}", &content_path, e))?;
 
+                    if !disable_cache {
+                        match disable_page_cache {
```
Comment (itowlson) [inline src/bartholomew.rs]: I think `if !disable_page_cache.unwrap_or_default()` would let you avoid the empty `Some(true)` case and the extra layer of nesting it inflicts on you.

File: src/bartholomew.rs [hunk 4]
```diff
@@ -114,10 +143,32 @@ pub fn render(req: Request) -> Result<Response> {
...
+                            Some(true) => {}
+                            _ => {
+                                let index_cache =
```
Comment (itowlson) [inline src/bartholomew.rs]: Is it practical to move this out to a named function e.g. `cache_rendered_page`?  The `render` function is already gargantuan and it might be good to minimise further growth...

File: src/content.rs [hunk 5]
```diff
@@ -124,40 +105,19 @@ pub fn all_pages(
...
             Ok(index_cache.contents)
         }
     }
 }
 
 pub struct IndexCache {
-    contents: BTreeMap<String, PageValues>,
-    cache_expiration: Option<DateTime<Utc>>,
+    pub contents: BTreeMap<String, PageValues>,
+    pub cache_expiration: DateTime<Utc>,
```
Comment (itowlson) [inline src/content.rs:137]: Is it safe to make this required - there won't be any existing caches that lack an expiration?

File: src/content.rs [hunk 6]
```diff
@@ -214,9 +174,13 @@ pub fn all_pages_load(dir: PathBuf, show_unpublished: bool) -> anyhow::Result<In
             }
         }
     }
+    let cache_expiration = match contains_unpublished {
+        true => earliest_unpublished.unwrap(),
+        false => DateTime::parse_from_rfc3339("2143-12-31T23:59:59+00:00")?.with_timezone(&Utc),
```
Comment (itowlson) [inline src/content.rs]: What is the significance of this string?  If you want something far in the future, could you use `UTC_MAX` rather than parsing an oddly specific string?

If the specific string is required, consider pulling it out to a constant with a meaningful name.

File: src/content.rs [hunk 7]
```diff
@@ -124,40 +105,19 @@ pub fn all_pages(
...
-                    // Cache expiry timer on the 1st line
-                    // The rest of the file contains the cache contents
-                    writeln!(f, "{cache_expiry_time}")?;
-                    write!(f, "{cache_data}")?;
-                }
-                _ => {
-                    eprintln!("Cannot create Cache file");
-                }
-            }
+            store.set("bartholomew_index_cache", cache_data)?;
```
Comment (itowlson) [inline src/content.rs]: Comment on line 109 (re 'files') might be out of date?

File: src/bartholomew.rs [hunk 8]
```diff
@@ -114,10 +143,32 @@ pub fn render(req: Request) -> Result<Response> {
...
+                                    all_pages_load(PathBuf::from_str(CONTENT_PATH)?, false)?;
+
+                                let rendereder_cache = RenderedPageCacheEntry {
```
Comment (itowlson) [inline src/bartholomew.rs]: tiny typo 
```suggestion
                                let rendered_cache = RenderedPageCacheEntry {
```

**Review by itowlson — COMMENTED** [review]
I found this a bit hard to follow, but I think you are in a bind - there is no real loader layer where you can inject caching behaviour, so you need to inject caching into existing large functions, and it ends up bringing lots of tactical considerations around things like time parsing right up into the cache and application layers.

Ideally it would be good to start unpicking bits of the structure, and adding traits and helper methods (e.g. `if cached.has_expired()`) rather than operating directly on strings and serialisation types.  But that would likely require a more significant refactor than you can fit in at this time.  Something to think about for the longer term maybe...

File: src/content.rs (hunk 5 above)
Comment (karthik2804) [reply src/content.rs:137]: This is safe as it is generated by going through the list of files (i.e) not read from somewhere. https://github.com/karthik2804/bartholomew/blob/4c53cd8e4453978ef84f15b5a07ade577063db5d/src/content.rs#L179

File: src/content.rs (hunk 6 above)
Comment (karthik2804) [reply src/content.rs]: The string is just a magic string formed from editing UTC_MAX. The parser did not seem happy to parse UTC_MAX as it was of the form `yyyyyy-mm-ddTHH:MM::SS:mmmmmm+00:00`. I will make sure again to check I was not missing something and if not I will pull it out with a constant.

File: src/content.rs (hunk 7 above)
Comment (karthik2804) [reply src/content.rs]: Yes - I will update the comment.

File: Cargo.toml (hunk 1 above)
Comment (karthik2804) [reply Cargo.toml]: Good point. I will tag it to `v0.9.0`

File: src/content.rs (hunk 2 above)
Comment (karthik2804) [reply src/content.rs]: Good point.

File: src/content.rs (hunk 6 above)
Comment (itowlson) [reply src/content.rs]: You wouldn't parse MAX_UTC (sorry messed up the name) would you?  It's already a DateTime.  https://docs.rs/chrono/latest/chrono/struct.DateTime.html#associatedconstant.MAX_UTC

File: src/bartholomew.rs (hunk 3 above)
Comment (karthik2804) [reply src/bartholomew.rs]: Thanks for pointing that out!

File: src/bartholomew.rs (hunk 4 above)
Comment (karthik2804) [reply src/bartholomew.rs]: Good point!

File: src/content.rs (hunk 6 above)
Comment (karthik2804) [reply src/content.rs]: This was just due to me parsing them incorrectly. Fixed it to use `MAX_UTC`.

**Comment by karthik2804** [comment]
@itowlson, I think I have addressed all the comments and I think this is ready for another round of review.

**Comment by karthik2804** [comment]
Not entirely sure how to solve the CLI error.

File: src/content.rs (hunk 6 above)
Comment (karthik2804) [reply src/content.rs]: I was trying to convert it to string and back. My mistake was I was converting it to a different format.

Comment (karthik2804) [reply src/content.rs]: I might actually have to take a look at this again tomorrow as the date format used content is "rfc339". I have to double-check a couple of things tomorrow.

**Comment by mikkelhegn** [comment]
Has this been tested in the cloud? (Not sure if we can), but I'm curious where the performance benefits come from? Moving from generating things as the request comes in, to read something pre-generated from the KV Store? Is that main thing?

**Comment by karthik2804** [comment]
We can't test it on the cloud yet. You are correct in the source of performance improvements, by keeping a copy of rendered stuff and using it until the cache is invalidated, the entire render cycle is skipped.

**Review by itowlson — APPROVED** [review]
LGTM!
